### PR TITLE
move res/ subdirectory to src/dmd/res/

### PR DESCRIPTION
**Pull request by WalterBright**
Builds should not reference files up and out of the build directory and to the left.

### CODE REVIEW

**Review by andralex — APPROVED** [review]

**Comment by WalterBright** [comment]
CyberShadow/DAutoTest says:
```
Invalid source/import path: /dev/shm/dtest/work/repo/dmd/res
```
where is that path specified?

**Comment by WalterBright** [comment]
ci/circleci:
```
Executing: /home/circleci/dmd/generated/linux/debug/64/dmd -conf= -m64 -oftest_results/dshell_prebuilt.o -c /home/circleci/dmd/test/tools/dshell_prebuilt/dshell_prebuilt.d -fPIC
/home/circleci/dmd/src/dmd/doc.d(364): Error: file `"default_ddoc_theme.ddoc"` cannot be found or not in a path specified with `-J`
```
where is this path specified in circleci?

**Comment by Geod24** [comment]
https://github.com/dlang/dmd/blob/ca636dcb8a56b53b8951ea8e5fca1bc581af639f/dub.sdl#L74

**Comment by WalterBright** [comment]
Thanks @Geod24 as `res` is not greppable given it appears everywhere. I never would have found it on my own.

**Comment by Geod24** [comment]
> Builds should not reference files up and out of the build directory and to the left.

Alternatively, I would say that build files should *not* be mixed with source files. Having makefiles and `build.d` in `src` has always irked me, but I think that's too late to change.

**Comment by WalterBright** [comment]
Thanks @Geod24 but there appears to be yet another one:
```
../dmd/generated/linux/release/64/dmd -J../dmd/res -J../dmd/generated/linux/release/64/ -c -o- -version=MARS -version=CoreDdoc \
	-version=StdDdoc -Df.generated/.prerelease-dummy.html \
	-Xf.generated/docs-prerelease.json -I../phobos @.generated/.prerelease-files.txt
../dmd/src/dmd/doc.d(364): Error: file `"default_ddoc_theme.ddoc"` cannot be found or not in a path specified with `-J`
```

**Comment by kubo39** [comment]
It seems unrelated to this failure, but I found one more case in ci.sh:test_dub_package.

```
"${build_path}/dmd" -version=NoBackend -version=GC -version=NoMain -Jgenerated/dub -Jres -Isrc -i -run test/dub_package/frontend.d
```

**Comment by WalterBright** [comment]
@kubo39 thanks, but where is that file? It's not in the dmd repository.

**Comment by kubo39** [comment]
@WalterBright 
here: https://github.com/dlang/dmd/blob/6ee04563a36ad64bbbc41e3aface6349402e3fab/ci.sh#L118

**Comment by WalterBright** [comment]
thanks, trying it now

**Comment by WalterBright** [comment]
same failure :-(

**Comment by kubo39** [comment]
It seems here:
https://github.com/dlang/dlang.org/blob/e0f9c04e9dd4e2e41207711aa798b897ce4c7ac6/posix.mak#L741

**Comment by WalterBright** [comment]
Thank you. Blocked by https://github.com/dlang/dlang.org/pull/2821

File: dub.sdl [hunk 1]
```diff
@@ -71,7 +71,7 @@ subPackage {
   name "frontend"
   targetType "library"
   sourcePaths "src/dmd"
-  stringImportPaths "res"
+  stringImportPaths "src/dmd/res"
```
Comment (wilzbach) [inline dub.sdl:74]: If you want to merge dlang.org PR without duplication, you need to temporarily add both directories here as dlang.org uses dmd as a library:

```
stringImportPaths "res"
stringImportPaths "src/dmd/res"
```

File: src/vcbuild/dmd.vcxproj [hunk 2]
```diff
@@ -122,7 +122,7 @@
     </Link>
     <DCompile>
       <VersionIdentifiers>MARS</VersionIdentifiers>
-      <StringImportPaths>..\..\res;$(OutDir)</StringImportPaths>
+      <StringImportPaths>../dmd/res;$(OutDir)</StringImportPaths>
```
Comment (wilzbach) [inline src/vcbuild/dmd.vcxproj:125]:
```suggestion
      <StringImportPaths>..\dmd\res;$(OutDir)</StringImportPaths>

```

(similar for the changes below)

This is the Visual Studio project file, so I am pretty sure the Windows slashes should be kept.

**Comment by wilzbach** [comment]
> Having makefiles and build.d in src has always irked me, but I think that's too late to change.

You're not only there. However, `build.d` isn't called directly by any CI (or packagers) yet, but only via the Makefiles.
So it would be possible to move build.d up to the root-level, adjust its paths and change the paths to build.d in the makefiles.
For full compatibility the `src/build.d` could print a deprecation message and execute sth. along: `$DC -run <scriptDir>/../build.d <args>`.

**Comment by MoonlightSentinel** [comment]
> You're not only there.

Ditto

> However, `build.d` isn't called directly by any CI (or packagers) yet, but only via the Makefiles.

https://github.com/dlang/dmd/blob/d263f806077b38be05bce8dc2dfb8230e4fc50e7/.github/workflows/runnable_cxx.yml#L274-L280

https://github.com/dlang/dmd/blob/d263f806077b38be05bce8dc2dfb8230e4fc50e7/.circleci/run.sh#L181-L186

**Comment by wilzbach** [comment]
> However, build.d isn't called directly by any CI (or packagers) yet, but only via the Makefiles.

Oh Thanks for the correction @MoonlightSentinel. I wasn't aware that we were already doing this, but as these files are still in the same repo + under our control, I guess it's a now or never decision, because I believe packagers will soon switch to use `src/build.d` directly and we want to make their version bumps as smooth as possible.

**Comment by wilzbach** [comment]
(restarted all CIs by force-pushing)

**Comment by wilzbach** [comment]
```

[13] [14] (/usr/share/texlive/texmf-dist/tex/latex/base/ts1cmtt.fd)
! You can't use `\spacefactor' in vertical mode.
\add@accent ...}\accent #1 #2\egroup \spacefactor 
                                                  \accent@spacefactor 
l.592     \t
            
? 
! Emergency stop.
\add@accent ...}\accent #1 #2\egroup \spacefactor 
                                                  \accent@spacefactor 
l.592     \t
            
!  ==> Fatal error occurred, no output PDF file produced!
Transcript written on .generated/dlangspec.log.
posix.mak:553: recipe for target '/dev/shm/dtest/work/repo/dlang.org/web/dlangspec.pdf' failed
```

@CyberShadow any ideas? I wasn't able to reproduce this locally. `dlangspec.tex` is the same for me with/without this PR.

**Comment by CyberShadow** [comment]
Oh, no. Not another PDF problem. :disappointed: 

I don't know what happened there. A random fluke?

**Comment by CyberShadow** [comment]
The cure: https://github.com/dlang/dlang.org/pull/2823

**Comment by WalterBright** [comment]
Thanks everyone for the help with this! There are more files like this that need fixing, but I'm a bit battle fatigued at the moment and will do them later.

**Pull request merged**